### DISCUSSION TITLE
How to implement 'afterPass()' & 'beforePass()'

### DISCUSSION THREAD

**Discussion by ccharlton** [GoogleChrome/lighthouse#15662] [Q&A] [answered]
I'm not seeing a straight path to get `afterPass()` and/or `beforePass()` to fire when using a custom Gatherer and Audit. I definitely see `getArtifact()` fire but from what I read, the hefty work should be done in `afterPass` versus `getArtifact`, if iirc.

**Comment by adamraine** [comment, marked as answer]
`afterPass` and `beforePass` were deprecated and then removed in favor of `startInstrumentation`, `stopInstrumentation` and `getArtifact` ([example](https://github.com/GoogleChrome/lighthouse/blob/main/docs/recipes/custom-audit/memory-gatherer.js#L18)).

What docs are you reading? Perhaps we need to update them.

**Comment by ccharlton** [comment]
I did a check, and you are correct that the current version of the docs don't reference afterPass and beforePass anymore. Legacy code examples may have tripped that wire.

Thank you for the clarification.
